### PR TITLE
ekf: correct calibration procedures + testing app introduction

### PR DESCRIPTION
**Pull request by niewim19**
JIRA: PD-43

### CODE REVIEW

File: ekf/kalman_implem.c [hunk 1]
```diff
@@ -133,7 +133,7 @@ static void init_state_vector(phmatrix_t *state)
 
 
 /* covariance matrox values inits */
-static void init_cov_vector(phmatrix_t *cov)
+static void init_cov_vector(phmatrix_t *cov, kalman_calib_t *calib)
```
Comment (MaciejPurski) [inline ekf/kalman_implem.c:136]: You are not using this parameter in this function.

File: ekf/measurement.c [hunk 2]
```diff
@@ -34,6 +30,16 @@
...
 #include "tools/phmatrix.h"
 
+#define EARTH_SEMI_MAJOR           6378137.0
+#define EARTH_SEMI_MINOR           6356752.3
+#define EARTH_ECCENTRICITY_SQUARED 0.006694384
+
+#define IMU_CALIB_AVG  2000
+#define BARO_CALIB_AVG 100
+
+struct {
```
Comment (MaciejPurski) [inline ekf/measurement.c:40]: Make it static.

File: ekf/measurement.c [hunk 3]
```diff
@@ -318,3 +351,15 @@ int acquireGpsMeasurement(vec_t *enu, vec_t *enu_speed, float *hdop)
 
 	return 0;
 }
+
+
+kalman_calib_t meas_calibGet(void)
+{
+	return meas_common.calib;
```
Comment (MaciejPurski) [inline ekf/measurement.c:358]: Don't return a structure as value - it causes unnecessary copying and increases stack usage. Return a pointer instead, or a const pointer if you don't want the client the content.

File: ekf/kalman_implem.c [hunk 4]
```diff
@@ -288,7 +288,7 @@ static void calcPredictionJacobian(phmatrix_t *F, phmatrix_t *state, time_t time
 
 
 /* initialization of prediction step matrix values */
-state_engine_t init_prediction_matrices(phmatrix_t *state, phmatrix_t *state_est, phmatrix_t *cov, phmatrix_t *cov_est, phmatrix_t *F, phmatrix_t *Q, time_t timeStep)
+state_engine_t init_prediction_matrices(phmatrix_t *state, phmatrix_t *state_est, phmatrix_t *cov, phmatrix_t *cov_est, phmatrix_t *F, phmatrix_t *Q, time_t timeStep, kalman_calib_t *calib)
```
[the remark this answers is not in the document]
Comment (niewim19) [reply ekf/kalman_implem.c:291]: Bigger refactor on this one was made. I will push it as separate commit.

File: ekf/measurement.c [hunk 5]
```diff
@@ -34,6 +30,16 @@
 #include "tools/rotas_dummy.h"
 #include "tools/phmatrix.h"
 
+#define EARTH_SEMI_MAJOR           6378137.0
+#define EARTH_SEMI_MINOR           6356752.3
+#define EARTH_ECCENTRICITY_SQUARED 0.006694384
```
Comment (agkaminski) [inline ekf/measurement.c:35]:
```suggestion
#define EARTH_SEMI_MAJOR           6378137.0F
#define EARTH_SEMI_MINOR           6356752.3F
#define EARTH_ECCENTRICITY_SQUARED 0.006694384F
```

**Review by agkaminski — COMMENTED** [review]
Empty file - ekf/measurement.h - is this intended?

File: ekf/measurement.c (hunk 5 above)
Comment (niewim19) [reply ekf/measurement.c:35]: No its there by a mistake. Thanks!

File: ekf/measurement.c [hunk 6]
```diff
@@ -186,99 +192,126 @@ static void ellipsoid_compensate(float *x, float *y, float *z, float *calib)
...
 
-		usleep(1000 * 5);
+	i = 0;
+	while (i < avg) {
+		if (sensc_imuGet(&accEvt, &gyrEvt, &magEvt) >= 0) {
+			meas_acc2si(&accEvt, &acc);
+			meas_gyr2si(&gyrEvt, &gyr);
+			meas_mag2si(&magEvt, &mag);
+
+			accAvg = vec_add(&accAvg, &acc);
```
Comment (MaciejPurski) [inline ekf/measurement.c:238]: You shouldn't return structures through stack.

File: ekf/tools/phmatrix.c [hunk 7]
```diff
@@ -43,6 +43,8 @@ int phx_assign(phmatrix_t * matrix, int m, int n, float * data)
 		matrix->transposed = 0;
 		return 0;
 	}
+
+	matrix->data = NULL;
```
Comment (MaciejPurski) [inline ekf/tools/phmatrix.c:47]: You are assigning `NULL` here, but what if it has already been set to dynamically allocated memory? I don't know how this function would be used, but I think you should check it and free the `data` pointer.

File: ekf/tools/phmatrix.c [hunk 8]
```diff
@@ -22,19 +22,19 @@ float * buf = NULL;
 unsigned int buflen = 0;
 
 /* dynamically allocate matrix */
```
Comment (MaciejPurski) [inline ekf/tools/phmatrix.c:24]: You are not in fact allocating a matrix here, but just the table of floats, but it is just a nitpick.

Comment (niewim19) [reply ekf/tools/phmatrix.c:24]: removed

File: ekf/measurement.c (hunk 6 above)
Comment (niewim19) [reply ekf/measurement.c:238]: As discussed, this needs huge refactor of math library. Nevertheless, that is true, not resolving this comment.

**Review by MaciejPurski — APPROVED** [review]

**Pull request merged**